zkvm/src/merkle.rs

```diff
@@ -0,0 +1,249 @@
+use merlin::Transcript;
+use subtle::ConstantTimeEq;
+
+use crate::errors::VMError;
+
+/// MerkleItem defines an item in the Merkle tree.
+pub trait MerkleItem: Sized {
+    /// Commits the hash of the item to Transcript.
+    fn commit(&self, t: &mut Transcript);
+}
+
+/// MerkleNeighbor is a step in a Merkle proof of inclusion.
+#[derive(Copy, Clone, PartialEq, Debug)]
+pub enum MerkleNeighbor {
+    Left([u8; 32]),
+    Right([u8; 32]),
+}
+
+/// Merkle tree of hashes with a given size.
+pub struct MerkleTree {
+    size: usize,
+    label: &'static [u8],
+    root: MerkleNode,
+}
+
+enum MerkleNode {
+    Leaf([u8; 32]),
+    Node([u8; 32], Box<MerkleNode>, Box<MerkleNode>),
+}
+
+impl MerkleTree {
+    /// Constructs a new MerkleTree based on the input list of entries.
+    pub fn build<M: MerkleItem>(label: &'static [u8], list: &[M]) -> Option<Self> {
+        if list.len() == 0 {
+            return None;
+        }
+        let t = Transcript::new(label);
+        Some(MerkleTree {
+            size: list.len(),
+            label,
+            root: Self::build_tree(t, list),
+        })
+    }
+
+    /// Builds a proof of inclusion for entry at the given index for the Merkle tree.
+    pub fn proof(&self, index: usize) -> Result<Vec<MerkleNeighbor>, VMError> {
```

Comment:
How about `path_to(i)` for this method and `verify_path` instead of `verify_proof`? So we use a more concrete term "path" instead of more abstract "proof" (because checking membership of all the items is also a proof, but not a path proof).